### ISSUE TITLE
IJavascriptCallback.ExecuteAsync never returns when callback no longer exists

### ISSUE THREAD

**Issue by gregmartinhtc** [cefsharp/CefSharp#1336]
I just upgraded to the latest version 43.0.0 of CefSharp (I believe the previous version was 41).
In 41 when an IJavascriptCallback went out of scope if you tried to call it the JavascriptResponse would be null which was my clue that it had gone out of scope and I would remove it from my callback list.

But in 43.0.0 if the IJavascriptCallback goes out of scope (i.e. the page changes) the call to IJavascriptCallback.ExecuteAsync never returns.  I haven't counted it but after somewhere between 10-20 never-returning calls to IJavascriptCallback.ExecuteAsync something crashes and the DevTools says its been disconnected and the browser becomes unresponsive.

I copied my current code below. I don't know if IJavascriptCallback.isDisposed was in 41, but I added a check for it to make sure I wasn't calling a disposed object and I'm not. As you can see I catch exceptions just in case also. In stepping through this code with the Visual Studio C# debugger I see that every call to ExecuteAsync never returns until something crashes.

Perhaps I've missed something added in 43.0.0 to handle this.  All I do know is it used to work very reliably in 41 returning a null JavascriptResponse when the callback was no longer valid.

Here's my code make the callback:

```c#
private async void CallJsDownloadCallback(IJavascriptCallback jsCallbackForDownloadStatus, int sinkCallbackId, DownloadManagerEventStatus eventType, Guid? itemGuid, string appId, ApplicationStatus appStatus, ApplicationInfo appInfo, long longValue, string stringValue) { JavascriptResponse response = null;
        try
        {
            if (!jsCallbackForDownloadStatus.IsDisposed)
            {
                string appInfoString = JsonConvert.SerializeObject(appInfo, defaultJsonSerializerSettings);

                // Here is the call that NEVER comes back if the IJavascriptCallback has went out of scope
                Task<JavascriptResponse> task = jsCallbackForDownloadStatus.ExecuteAsync(sinkCallbackId, (int)eventType, itemGuid?.ToString(), appId, (int)appStatus, appInfoString, longValue, stringValue);

                response = await task;
            }
            else
            {
                Debug.WriteLine("jsCallbackForStoreStatus was disposed");
            }
        }
        catch (Exception ex)
        {
            Debug.WriteLine("CallJsDownloadCallback exception: " + ex.Message);
        }

        if (response == null)
        {
            // Assume the Javascript callback is now invalid
            m_managerServices.DownloadManagerServiceEventSink.UnsubscribeDownloadEvents(sinkCallbackId);
        }
    }
```

**Comment by amaitland** [comment]
> Perhaps I've missed something added in 43.0.0 to handle this. All I do know is it used to work very reliably in 41 returning a null JavascriptResponse when the callback was no longer valid.

`Callbacks` were rewritten to use the `CEF native IPC`, previously they used `WCF` (which is still used for some parts the `JSB` implementation). Without looking at this in too much detail, I'm guessing we need to kill off all pending `callbacks` in `OnBeforeClose`

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/Internals/ClientAdapter.cpp#L206

Are you able to contribute a `PR` that reproduces the problem? Make it easier to validate a possible solution.

**Comment by gregmartinhtc** [comment]
Previously I just installed the NuGet packages so I'm a neophyte on how to provide feedback so please bear with me.  I did clone the repo and used the master branch to reproduce the issue easily.

To reproduce the issue just open the VS project and go to `CefSharp.Example\Resources\BindingTest.html` and edit the JavaScript Callback Test by adding a call to "**`window.location.assign("http://www.google.com");`**" right before the `testCallback()` method returns as follows:

```javascript
        <p>
            Javscript Callback Test
            <br />
            <script type="text/javascript">
                function callback(s)
                {
                    var result = document.getElementById('cbresult');
                    result.innerText += "Callback: " + s + "" + Date();
                }

                function testCallback()
                {
                    bound.testCallback(callback);

                    var result = document.getElementById('cbresult');
                    result.innerText = "The function has returned: " + Date() + "\n";

                    window.location.assign("http://www.google.com");
                }
            </script>
            <button onclick="testCallback()">Test Callback</button>
            <br />
            <span id="cbresult"></span>
        </p>
```

I used the WinForms.Example app to repro the issue by running it and navigating to "**custom://cefsharp/BindingTest.html**" and clicking on the "Test Callback" button.  After clicking the button if you wait about 10 seconds after it goes to the google page the renderer process will crash.

If you set a breakpoint in `CefSharp.Example\BoundObject.cs::TestCallback(IJavascriptCallback javascriptCallback)` method in the call to "**`await javascriptCallback.ExecuteAsync("This callback from C# was delayed " + taskDelay + "ms");`**" you'll see that it never returns before or after the renderer crash.

I do see in `CefSharp.Core\JavascriptCallbackProxy.cpp:JavascriptCallbackProxy::ExecuteAsync(cli::array<Object^>^ parameters)` the call to "`browserWrapper->SendProcessMessage(CefProcessId::PID_RENDERER, callbackMessage);`" returns, but I suspect internally that is the call the ends up causing the crash in the renderer process.

I don't quite know the difference between `RegisterJsObject()` and `RegisterAsyncJsObject()` and I haven't had time yet to figure that out (I thought maybe one used `WCF` and the other used `CEF native IPC` as you spoke about), but whatever the difference in this case the BoundObject is registered with `RegisterJsObject()`.  In my real project I found that whether I used `RegisterJsObject()` or `RegisterAsyncJsObject()` the same crash occurred.

I suspect your explanation about the need to kill off pending callbacks is probably just half the solution for callbacks made while the JavaScript callback exists, but not processed until after the callback as been disposed of.  In this above example, though, it would seem to me the CefSharp.Core\JavascriptCallbackProxy.cpp:JavascriptCallbackProxy should have been disposed of when it went out of scope (when the URL page was changed) in which case the call to `DisposedGuard()` in `JavascriptCallbackProxy::ExecuteAsync(...)` would have thrown an exception and the SendProcessMessage would never have been made.

I'm going to look at little further into this, but as a neophyte to the project's source code I'm swimming through it mostly blind and making educated guesses as to what things mean.

**Comment by amaitland** [comment]
> To reproduce the issue just open the VS project and go to CefSharp.Example\Resources\BindingTest.html and edit the JavaScript Callback Test by adding a call to "window.location.assign("http://www.google.com");" right before the testCallback() method returns as follows:

Can you submit a `PR` that includes these changes please. Instead of changing `BindingTest.html` can you create a new test to demo the problem, see #1318 for an example.

> I don't quite know the difference between RegisterJsObject() and RegisterAsyncJsObject() and I haven't had time yet to figure that out (I thought maybe one used WCF and the other used CEF native IPC as you spoke about),

Correct, the `Async` version uses the `Native IPC`, it also only works for methods, uses javascript promises to complete and runs in an async fashion where the regular version blocks the render process. The `Async` version is new to the `43.0.0` release.

> I suspect your explanation about the need to kill off pending callbacks is probably just half the solution for callbacks made while the JavaScript callback exists, but not processed until after the callback as been disposed of. In this above example, though, it would seem to me the CefSharp.Core\JavascriptCallbackProxy.cpp:JavascriptCallbackProxy should have been disposed of when it went out of scope (when the URL page was changed) in which case the call to DisposedGuard() in JavascriptCallbackProxy::ExecuteAsync(...) would have thrown an exception and the SendProcessMessage would never have been made.

`JavascriptCallbackProxy` should be disposed as well. If the call to `SendProcessMessage ` crashes the render process then there's likely another small problem which should also be fixed. We're using `process-per`tab` so a single render process is used for a `ChromiumWebBrowser` instance.

**Comment by gregmartinhtc** [comment]
Ah, now I understand what you mean by a PR.  A Pull Request with new test.  Ah, ok.  I'll do so a little later.  I'm digging down deeper for a bit trying to understand something.

**Comment by amaitland** [comment]
> A Pull Request with new test.

Yes please :+1:

**Comment by gregmartinhtc** [comment]
I have fixed the issue, but, again, I'm a newbie to this project (and actually to using GitHub on non-private projects) and I'm confused.  Do you want me to provide a fix/pull request on the main or the 43 branch?

**Comment by gregmartinhtc** [comment]
I also assume I need to be granted some privileges to push my changes?  I created a branch off of main, updated it, and then tried to push the branch, but was forbidden.  What is the process for submitting my code fix and new test?

**Comment by amaitland** [comment]
> Do you want me to provide a fix/pull request on the main or the 43 branch?

Either is fine in this case.

> I also assume I need to be granted some privileges to push my changes? I created a branch off of main, updated it, and then tried to push the branch, but was forbidden. What is the process for submitting my code fix and new test?

If you fork `CefSharp`, push changes to your fork, then create a `PR`, changes can then be reviewed before `merge`.

**Comment by gregmartinhtc** [comment]
Thanks!  Like I said I'm a neophyte about contributing to public projects
on Github.  All of my previous experience was with private corporate
projects on Github where we just created branches and created pull requests
from them.  I did look briefly into forking and actually created a fork on
Github, but I was struggling with how to pull down the fork into Visual
Studio.  I'm on my phone right now so I'll have to wait until tomorrow to
figure it out, but thanks so much for pointing me in the right direction.
:-)
